// ntcontrol-connection/dist/Responses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResponseDescription = exports.ResponseCode = void 0;
var ResponseCode;
(function (ResponseCode) {
    ResponseCode["ERR1"] = "ERR1";
    ResponseCode["ERR2"] = "ERR2";
    ResponseCode["ERR3"] = "ERR3";
    ResponseCode["ERR4"] = "ERR4";
    ResponseCode["ERR5"] = "ERR5";
    ResponseCode["ERRA"] = "ERRA";
    ResponseCode["ER401"] = "ER401";
})(ResponseCode = exports.ResponseCode || (exports.ResponseCode = {}));
function getResponseDescription(code) {
    switch (code) {
        case ResponseCode.ERR1: return 'Undefined control command';
        case ResponseCode.ERR2: return 'Parameter out of range';
        case ResponseCode.ERR3: return 'Busy state or unavailable period';
        case ResponseCode.ERR4: return 'Time out or unavailable period';
        case ResponseCode.ERR5: return 'Invalid data length';
        case ResponseCode.ERRA: return 'Mismatching state of a password';
        case ResponseCode.ER401: return 'Error occurred on processing command';
    }
    return undefined;
}
exports.getResponseDescription = getResponseDescription;
