// ntcontrol-connection/dist/Types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisplayGridLines = exports.CommandType = exports.EdgeBlending = exports.CustomMasking = exports.ActionSpeed = exports.TestPattern = exports.LampStatus = exports.LampControlStatus = exports.LensMemory = exports.NoSignalShutOff = exports.ShutterFade = exports.ScreenSetting = exports.ColorMatching = exports.Aspect = exports.Geometry = exports.ColorTemperature = exports.PictureMode = exports.ProjectorInput = void 0;
var ProjectorInput;
(function (ProjectorInput) {
    ProjectorInput["COMPUTER1"] = "RG1";
    ProjectorInput["COMPUTER2"] = "RG2";
    ProjectorInput["VIDEO"] = "VID";
    ProjectorInput["Y/C"] = "SVD";
    ProjectorInput["DVI"] = "DVI";
    ProjectorInput["HDMI1"] = "HD1";
    ProjectorInput["HDMI2"] = "HD2";
    ProjectorInput["NETWORK/USB"] = "NWP";
    ProjectorInput["Panasonic Application"] = "PA1";
    ProjectorInput["Miracast/Mirroring"] = "MC1";
    ProjectorInput["MEMORY VIEWER"] = "MV1";
    ProjectorInput["SDI1"] = "SD1";
    ProjectorInput["SDI2"] = "SD2";
    ProjectorInput["DIGITAL LINK"] = "DL1";
})(ProjectorInput = exports.ProjectorInput || (exports.ProjectorInput = {}));
var PictureMode;
(function (PictureMode) {
    PictureMode["DYNAMIC"] = "DYN";
    PictureMode["NATURAL"] = "NAT";
    PictureMode["STANDARD"] = "STD";
    PictureMode["BLACK BOARD"] = "BBD";
    PictureMode["WHITE BOARD"] = "WBD";
    PictureMode["CINEMA"] = "CIN";
    PictureMode["GRAPHIC"] = "GRA";
    PictureMode["DICOM SIM"] = "DIC";
    PictureMode["USER"] = "USR";
})(PictureMode = exports.PictureMode || (exports.PictureMode = {}));
var ColorTemperature;
(function (ColorTemperature) {
    ColorTemperature["LOW"] = "0";
    ColorTemperature["DEFAULT"] = "1";
    ColorTemperature["HIGH"] = "2";
    ColorTemperature["USER1"] = "04";
    ColorTemperature["USER2"] = "09";
    ColorTemperature["3200K"] = "3200";
    ColorTemperature["3300K"] = "3300";
    ColorTemperature["9200K"] = "9200";
    ColorTemperature["9300K"] = "9300";
})(ColorTemperature = exports.ColorTemperature || (exports.ColorTemperature = {}));
var Geometry;
(function (Geometry) {
    Geometry["OFF"] = "+00000";
    Geometry["KEYSTONE"] = "+00001";
    Geometry["CURVED"] = "+00002";
    Geometry["PC-1"] = "+00003";
    Geometry["PC-2"] = "+00004";
    Geometry["PC-3"] = "+00005";
    Geometry["CORNER-CORRECTION"] = "+00010";
})(Geometry = exports.Geometry || (exports.Geometry = {}));
var Aspect;
(function (Aspect) {
    Aspect["AUTO/VID AUTO/DEFAULT"] = "0";
    Aspect["NORMAL(4:3)"] = "1";
    Aspect["WIDE(16:9)"] = "2";
    Aspect["NATIVE(through)"] = "5";
    Aspect["FULL(HV FIT)"] = "6";
    Aspect["H-FIT"] = "9";
    Aspect["V-FIT"] = "10";
})(Aspect = exports.Aspect || (exports.Aspect = {}));
var ColorMatching;
(function (ColorMatching) {
    ColorMatching["OFF"] = "+00000";
    ColorMatching["3COLORS"] = "+00001";
    ColorMatching["7COLORS"] = "+00002";
    ColorMatching["MEASURED"] = "+00004";
})(ColorMatching = exports.ColorMatching || (exports.ColorMatching = {}));
var ScreenSetting;
(function (ScreenSetting) {
    ScreenSetting["16:10"] = "0";
    ScreenSetting["16:9"] = "1";
    ScreenSetting["4:3"] = "2";
})(ScreenSetting = exports.ScreenSetting || (exports.ScreenSetting = {}));
var ShutterFade;
(function (ShutterFade) {
    ShutterFade["0.0s(OFF)"] = "0.0";
    ShutterFade["0.5s"] = "0.5";
    ShutterFade["1.0s"] = "1.0";
    ShutterFade["1.5s"] = "1.5";
    ShutterFade["2.0s"] = "2.0";
    ShutterFade["2.5s"] = "2.5";
    ShutterFade["3.0s"] = "3.0";
    ShutterFade["3.5s"] = "3.5";
    ShutterFade["4.0s"] = "4.0";
    ShutterFade["5.0s"] = "5.0";
    ShutterFade["7.0s"] = "7.0";
    ShutterFade["10.0s"] = "10.0";
})(ShutterFade = exports.ShutterFade || (exports.ShutterFade = {}));
var NoSignalShutOff;
(function (NoSignalShutOff) {
    NoSignalShutOff["DISABLE"] = "00";
    NoSignalShutOff["10min"] = "10";
    NoSignalShutOff["20min"] = "20";
    NoSignalShutOff["30min"] = "30";
    NoSignalShutOff["40min"] = "40";
    NoSignalShutOff["50min"] = "50";
    NoSignalShutOff["60min"] = "60";
    NoSignalShutOff["70min"] = "70";
    NoSignalShutOff["80min"] = "80";
    NoSignalShutOff["90min"] = "90";
})(NoSignalShutOff = exports.NoSignalShutOff || (exports.NoSignalShutOff = {}));
var LensMemory;
(function (LensMemory) {
    LensMemory["LENS MEMORY1"] = "+00000";
    LensMemory["LENS MEMORY2"] = "+00001";
    LensMemory["LENS MEMORY3"] = "+00002";
    LensMemory["LENS MEMORY4"] = "+00003";
    LensMemory["LENS MEMORY5"] = "+00004";
    LensMemory["LENS MEMORY6"] = "+00005";
    LensMemory["LENS MEMORY7"] = "+00006";
    LensMemory["LENS MEMORY8"] = "+00007";
    LensMemory["LENS MEMORY9"] = "+00008";
    LensMemory["LENS MEMORY10"] = "+00009";
})(LensMemory = exports.LensMemory || (exports.LensMemory = {}));
var LampControlStatus;
(function (LampControlStatus) {
    LampControlStatus["LAMP OFF"] = "0";
    LampControlStatus["In turning ON"] = "1";
    LampControlStatus["LAMP ON"] = "2";
    LampControlStatus["LAMP Cooling"] = "3";
})(LampControlStatus = exports.LampControlStatus || (exports.LampControlStatus = {}));
var LampStatus;
(function (LampStatus) {
    LampStatus["ALL OFF"] = "0";
    LampStatus["ALL ON"] = "1";
    LampStatus["1:ON, 4:ON"] = "2";
    LampStatus["2:ON, 3:ON"] = "3";
    LampStatus["1:ON, 2:ON, 3:ON"] = "4";
    LampStatus["1:ON, 2:ON, 4:ON"] = "5";
    LampStatus["1:ON, 3:ON, 4:ON"] = "6";
    LampStatus["2:ON, 3:ON, 4:ON"] = "7";
    LampStatus["1:ON"] = "8";
    LampStatus["2:ON"] = "9";
    LampStatus["3:ON"] = "10";
    LampStatus["4:ON"] = "11";
})(LampStatus = exports.LampStatus || (exports.LampStatus = {}));
var TestPattern;
(function (TestPattern) {
    TestPattern["Off"] = "00";
    TestPattern["White"] = "01";
    TestPattern["Black"] = "02";
    TestPattern["Red"] = "22";
    TestPattern["Green"] = "23";
    TestPattern["Blue"] = "24";
    TestPattern["Cyan"] = "28";
    TestPattern["Magenta"] = "29";
    TestPattern["Yellow"] = "30";
    TestPattern["Window"] = "05";
    TestPattern["Window (Inversion)"] = "06";
    TestPattern["Colorbars (Vertical)"] = "08";
    TestPattern["Colorbars (Horizontal)"] = "51";
    TestPattern["16:9 / 4:3"] = "59";
    TestPattern["Crosshatch (White)"] = "07";
    TestPattern["Crosshatch (Red)"] = "70";
    TestPattern["Crosshatch (Green)"] = "71";
    TestPattern["Crosshatch (Blue)"] = "72";
    TestPattern["Crosshatch (Cyan)"] = "73";
    TestPattern["Crosshatch (Magenta)"] = "74";
    TestPattern["Crosshatch (Yellow)"] = "75";
    TestPattern["Flag"] = "03";
    TestPattern["Flag (Inversion)"] = "04";
    TestPattern["Lamp"] = "09";
    TestPattern["Convergence"] = "11";
    TestPattern["10%-Luminance"] = "25";
    TestPattern["5%-Luminance"] = "26";
    TestPattern["3D-1"] = "80";
    TestPattern["3D-2"] = "81";
    TestPattern["3D-3"] = "82";
    TestPattern["3D-4"] = "83";
})(TestPattern = exports.TestPattern || (exports.TestPattern = {}));
var ActionSpeed;
(function (ActionSpeed) {
    ActionSpeed["SLOW+"] = "+00000";
    ActionSpeed["SLOW-"] = "+00001";
    ActionSpeed["NORMAL+"] = "+00100";
    ActionSpeed["NORMAL-"] = "+00101";
    ActionSpeed["FAST+"] = "+00200";
    ActionSpeed["FAST-"] = "+00201";
})(ActionSpeed = exports.ActionSpeed || (exports.ActionSpeed = {}));
var CustomMasking;
(function (CustomMasking) {
    CustomMasking["OFF"] = "+00000";
    CustomMasking["PC-1"] = "+00001";
    CustomMasking["PC-2"] = "+00002";
    CustomMasking["PC-3"] = "+00003";
})(CustomMasking = exports.CustomMasking || (exports.CustomMasking = {}));
var EdgeBlending;
(function (EdgeBlending) {
    EdgeBlending["OFF"] = "+00000";
    EdgeBlending["ON"] = "+00001";
    EdgeBlending["USER"] = "+00002";
})(EdgeBlending = exports.EdgeBlending || (exports.EdgeBlending = {}));
var CommandType;
(function (CommandType) {
    CommandType[CommandType["Ascii"] = 0] = "Ascii";
    CommandType[CommandType["Binary"] = 1] = "Binary";
})(CommandType = exports.CommandType || (exports.CommandType = {}));
var DisplayGridLines;
(function (DisplayGridLines) {
    DisplayGridLines["OFF"] = "FFFFFF";
    DisplayGridLines["White"] = "FFFFFF";
    DisplayGridLines["Black"] = "000000";
    DisplayGridLines["Red"] = "FF0000";
    DisplayGridLines["Green"] = "00FF00";
    DisplayGridLines["Blue"] = "0000FF";
    DisplayGridLines["Cyan"] = "00FFFF";
    DisplayGridLines["Magenta"] = "FF00FF";
    DisplayGridLines["Yellow"] = "FFFF00";
})(DisplayGridLines = exports.DisplayGridLines || (exports.DisplayGridLines = {}));
